pkg/operator/controller/dns/controller.go

```diff
@@ -0,0 +1,178 @@
+	dnsConfig := &configv1.DNS{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dns 'cluster': %v", err)
+	}
+
+	record := &iov1.DNSRecord{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, record); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("dnsrecord not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		log.Error(err, "failed to get dnsrecord; will retry", "dnsrecord", request.NamespacedName)
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	// If the DNS record was deleted, clean up and return.
+	if record.DeletionTimestamp != nil {
+		if err := r.delete(record); err != nil {
+			log.Error(err, "failed to delete dnsrecord; will retry", "dnsrecord", record)
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+		}
+		return reconcile.Result{}, nil
+	}
+
+	ingressName, ok := record.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		log.V(2).Info("warning: dnsrecord is missing owner label", "dnsrecord", record, "ingresscontroller", ingressName)
+		return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+	}
+	ingress := &operatorv1.IngressController{}
+	if err := r.cache.Get(context.TODO(), types.NamespacedName{Namespace: record.Namespace, Name: ingressName}, ingress); err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: what should we do here? something upstream could detect and delete the orphan? add new conditions?
+			// is it safe to retry without verifying the owner isn't a new object?
+			log.V(2).Info("warning: dnsrecord owner not found; the dnsrecord may be orphaned; will retry", "dnsrecord", record, "ingresscontroller", ingressName)
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		} else {
+			log.Error(err, "failed to get ingresscontroller for dnsrecord; will retry", "dnsrecord", record, "ingresscontroller", ingressName)
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+		}
+	}
+
+	// TODO: Will soon need to honor scope of the ingresscontroller's LoadBalancer
+	// strategy.
+	zones := []configv1.DNSZone{}
+	if dnsConfig.Spec.PrivateZone != nil {
+		zones = append(zones, *dnsConfig.Spec.PrivateZone)
+	}
+	if dnsConfig.Spec.PublicZone != nil {
+		zones = append(zones, *dnsConfig.Spec.PublicZone)
+	}
+
+	// TODO: If we compare desired zones with zones from status (i.e. actual
+	// zones), we can filter to list to those zones which haven't already been
+	// successfully ensured. Then the DNS provider wouldn't need such aggressive
+	// internal caching.
+	result := reconcile.Result{}
+	statuses := []iov1.DNSZoneStatus{}
+	for i := range zones {
+		zone := zones[i]
+		conditions := []iov1.DNSZoneCondition{}
+		err := r.dnsProvider.Ensure(record, zone)
+		if err != nil {
+			conditions = append(conditions, iov1.DNSZoneCondition{
```

Comment:
Didn't we decide to set `Failed=False` when `err` is nil?

Reply:
I don't think we explicitly decided. If we publish True when there's a problem, does that imply we must publish False when there's not a problem? There's no internal reason we need the condition when things are normal, not sure what the conventions are here. We can change in a followup.